fw/http_stream_sched.c

```diff
@@ -0,0 +1,566 @@
+/**
+ *		Tempesta FW
+ *
+ * Copyright (C) 2023 Tempesta Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+#include "http_stream_sched.h"
+#include "http_stream.h"
+#include "connection.h"
+
+static inline void
+tfw_h2_stream_sched_spin_lock_assert(TfwStreamSched *sched)
+{
+	TfwH2Ctx *ctx = container_of(sched, TfwH2Ctx, sched);
+	TfwH2Conn *conn = container_of(ctx, TfwH2Conn, h2);
+
+	/*
+	 * All scheduler functions schould be called under the
+	 * socket lock.
+	 */
+	assert_spin_locked(&((TfwConn *)conn)->sk->sk_lock.slock);
+}
+
+/**
+ * Remove stream from the ebtree of the blocked streams and insert
+ * it in the ebtree of active streams. Should be called only for
+ * active streams or the streams with active children, which is
+ * not already in the ebtree of active streams.
+ */
+static void
+tfw_h2_stream_sched_insert_active(TfwStream *stream, u64 deficit)
+{
+	TfwStreamSchedEntry *parent = stream->sched.parent;
+
+	BUG_ON(!parent || (!tfw_h2_stream_is_active(stream) &&
+	       !tfw_h2_stream_sched_is_active(&stream->sched)));
+	BUG_ON(stream->sched_state == HTTP2_STREAM_SCHED_STATE_ACTIVE);
+
+	eb64_delete(&stream->sched_node);
+	stream->sched_node.key = deficit;
+	eb64_insert(&parent->active, &stream->sched_node);
```

Comment:
**IMPORTANT** These delete & insert aren't nice at all, especially with ebtree node size about 48 bytes. We need a tree to keep the streams sorted by deficit and we use ebtree because it wins for small trees. Did you consider just to keep an array of unsorted items (deficit keys, preferably decoupled from the rest of the data structures) and use SIMD for fast selection?  Something like https://www.youtube.com/watch?v=IMnbytvHCjM

Probably, we can merge this PR with ebtree, but it seems we need another PR for performance optimization of this work and this can be done there.

Reply:
There is a problem here. We have two trees. First tree is a tree of stream dependency. The second tree is a tree of priority on each level. For example if we can have such tree:
                              1
                    3                  5
             7  9  11  13     15    17
In this case when we choose the most priority stream we first of all try stream 1, if it is not active we try to choose the most priority stream between 3 and 5, if 3 is not active but have more priority then 5 and have active children we  should choose the most priority stream from 7 9 11 13. More over we should remove not only chosen stream but also it's parent , recalculate there deficits and insert them in the tree again. So SIMD is not good for this purpose if we want to comply with the RFC

Reply:
So I don't move this comment to 1411, if you don't agree with me we need to discuss it again

Reply:
Yeah, we discussed on the call that we need not only sorted list of streams, but also manage dependency between the streams, so my proposal doesn't make much sense and we really need a tree